### PR TITLE
[DC-1289]/[DC-1290] Create deid tier script that reads and validated command line parameters

### PR DESCRIPTION
**Pull request by amcgrenera-vumc**

### CODE REVIEW

File: data_steward/tools/create_tier.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        '--release_tag',
+        action='store',
+        dest='release_tag',
+        help='release tag for dataset in the format of YYYYq#r#',
+        required=True)
+    parser.add_argument('-d',
+                        '--deid_stage',
+                        action='store',
+                        dest='deid_stage',
+                        help='deid stage (deid, base or clean)',
```
Comment (lrwb-aou) [inline data_steward/tools/create_tier.py:113]: I think you should try using the argparse [`type`](https://docs.python.org/3/library/argparse.html#type) or [`choices`](https://docs.python.org/3/library/argparse.html#choices) arguments.

File: data_steward/tools/create_tier.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+                        action='store',
+                        dest='project_id',
+                        help=('Project associated with the '
+                              'input dataset.'),
+                        required=True)
+    parser.add_argument('-t',
+                        '--tier',
+                        action='store',
+                        dest='tier',
+                        help='controlled or registered tier',
```
Comment (lrwb-aou) [inline data_steward/tools/create_tier.py:94]: I think you should try using the argparse [`type`](https://docs.python.org/3/library/argparse.html#type) or [`choices`](https://docs.python.org/3/library/argparse.html#choices) arguments.

File: data_steward/tools/create_tier.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                        action='store',
+                        dest='idataset',
+                        help='Name of the input dataset',
+                        required=True)
+    parser.add_argument(
+        '-r',
+        '--release_tag',
+        action='store',
+        dest='release_tag',
+        help='release tag for dataset in the format of YYYYq#r#',
```
Comment (lrwb-aou) [inline data_steward/tools/create_tier.py:107]: I think you should try using the argparse [`type`](https://docs.python.org/3/library/argparse.html#type)  argument.

File: data_steward/tools/create_tier.py (hunk 2 above)
Comment (amcgrenera-vumc) [reply data_steward/tools/create_tier.py:94]: Oh neat, I haven't used those before. I will adjust the code and use a user defined function for they type for `release_tag` and either choice or type for the other two. Thanks!

Comment (lrwb-aou) [reply data_steward/tools/create_tier.py:94]: Just a quick review of your code does make it look like this PR is covering both DC-1289 and DC-1290.

File: data_steward/tools/create_tier.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    :param arg_value: release tag parameter passed through either the command line arguments
+    :return: arg_value
+    """
+
+    release_tag_regex = re.compile(r'[0-9]{4}q[0-9]r[0-9]')
+    if not re.match(release_tag_regex, arg_value):
+        LOGGER.error(
+            f"Parameter ERROR {arg_value} is in an incorrect format, accepted: YYYYq#r#"
+        )
+    return arg_value
```
Comment (lrwb-aou) [inline data_steward/tools/create_tier.py]: very nice!

File: data_steward/tools/create_tier.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+
+def create_tier(credentials_filepath, project_id, tier, input_dataset,
+                release_tag, deid_stage):
+    """
+    This function is hte main entry point for the deid process.
```
Comment (lrwb-aou) [inline data_steward/tools/create_tier.py:59]: nit:  hte should be the.

File: data_steward/tools/create_tier.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+
+def validate_tier_param(tier):
+    """
+    helper function to validate the tier parameter passed is either 'controlled' or 'registered'
+
+    :param tier: tier parameter passed through from either a list or command line argument
+    :return: nothing, breaks if not valid
+    """
+    if tier.lower() not in ['controlled', 'registered']:
```
Comment (lrwb-aou) [inline data_steward/tools/create_tier.py:20]: Might make this a module level list so you can reference it directly in the argument parser setup and here.  Then you only need to update it in one place.

```suggestion
    if tier.lower() not in TIER_LIST:
```
where `TIER_LIST = ['controlled', 'registered']`

File: data_steward/tools/create_tier.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
+# Python imports
+import argparse
+import logging
+import re
+
+# Third party imports
+
+# Project imports
+
```
Comment (lrwb-aou) [inline data_steward/tools/create_tier.py:9]: Use the pipeline logging setup utility at `data_steward/utils/pipeline_logging.py`.

See usage info in `configure(level=logging.INFO, add_console_handler=False)`.

Just add it the configure call to your main function.

File: data_steward/tools/create_tier.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+def validate_deid_stage_param(deid_stage):
+    """
+    helper function to validate the deid_stage parameter passed is correct, must be 'deid', 'base' or 'clean'
+
+    :param deid_stage: deid_stage parameter passed through from either a list or command line argument
+    :return: nothing, breaks if not valid
+    """
+    if deid_stage not in ['deid', 'base', 'clean']:
```
Comment (lrwb-aou) [inline data_steward/tools/create_tier.py:33]: Consider moving this list a a module level list so you can reuse it in the argument parser setup.

```suggestion
    if deid_stage not in DEID_STAGE_LIST:
```

where `DEID_STAGE_LIST = ['deid', 'base', 'clean']`.

File: data_steward/tools/create_tier.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+                        help=('Project associated with the '
+                              'input dataset.'),
+                        required=True)
+    parser.add_argument('-t',
+                        '--tier',
+                        action='store',
+                        dest='tier',
+                        help='controlled or registered tier',
+                        required=True,
+                        choices=['controlled', 'registered'])
```
Comment (lrwb-aou) [inline data_steward/tools/create_tier.py:98]: If using common  list, can set as: `choices = TIER_LIST`.

File: data_steward/tools/create_tier.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+        help='release tag for dataset in the format of YYYYq#r#',
+        required=True,
+        type=valid_release_tag)
+    parser.add_argument('-d',
+                        '--deid_stage',
+                        action='store',
+                        dest='deid_stage',
+                        help='deid stage (deid, base or clean)',
+                        required=True,
+                        choices=['deid', 'base', 'clean'])
```
Comment (lrwb-aou) [inline data_steward/tools/create_tier.py:119]: If using common list, could change to `choices=DEID_STAGE_LIST`.

File: data_steward/tools/create_tier.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+    return parser.parse_args(args)
+
+
+def main(raw_args=None):
+    args = parse_deid_args(raw_args)
+    create_tier(args.credentials_filepath, args.project_id, args.tier,
```
Comment (lrwb-aou) [inline data_steward/tools/create_tier.py:125]: Probably add the piece about configuring the pipeline here, after parsing the arguments.  To make it dynamic, you may want to add an argument allowing the logs to print to the console, and pass that to the configure function.

File: data_steward/tools/create_tier.py [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+    :param tier: controlled or registered tier intended for the output dataset
+    :param input_dataset: name of the input dataset
+    :param release_tag: release tag for dataset in the format of YYYYq#r#
+    :param deid_stage: deid stage (deid, base or clean)
+    :return: name of created controlled or registered dataset
+    """
+
+    # validation of params
+    validate_tier_param(tier)
+    validate_deid_stage_param(deid_stage)
```
Comment (lrwb-aou) [inline data_steward/tools/create_tier.py:73]: If you are trying to get DC-1290 done too, you should call the `valid_release_tag` function here too.

**Review by lrwb-aou — COMMENTED** [review]
Looking good!  I think there were some logging improvements that were put in place while you were on leave.  Let me know if you'd like an overview on them.  Most of the suggestions are just to improve maintainability.

**Comment by amcgrenera-vumc** [comment]
@lrwb-aou this is now revised, please take a look when you get a moment. Thanks!

File: data_steward/tools/create_tier.py (hunk 7 above)
Comment (amcgrenera-vumc) [reply data_steward/tools/create_tier.py:10]: added!

File: data_steward/tools/create_tier.py (hunk 6 above)
Comment (amcgrenera-vumc) [reply data_steward/tools/create_tier.py:21]: Good thinking, thank you!

File: data_steward/tools/create_tier.py (hunk 12 above)
Comment (amcgrenera-vumc) [reply data_steward/tools/create_tier.py:77]: added!

File: data_steward/tools/create_tier.py (hunk 11 above)
Comment (amcgrenera-vumc) [reply data_steward/tools/create_tier.py:132]: Let me know if I did this correctly.

File: data_steward/tools/create_tier.py [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+                        help='deid stage (deid, base or clean)',
+                        required=True,
+                        choices=DEID_STAGE_LIST)
+    return parser.parse_args(args)
+
+
+def main(raw_args=None):
+    args = parse_deid_args(raw_args)
+    pipeline_logging.configure(level=logging.DEBUG,
+                               add_console_handler=args.console_log)
```
Comment (lrwb-aou) [inline data_steward/tools/create_tier.py:131]: The only issue I see is `args.console_log` has not been defined by the parser in `parse_deid_args`.

Comment (amcgrenera-vumc) [reply data_steward/tools/create_tier.py:131]: ah yes forgot to add that in, adding that now.

**Comment by amcgrenera-vumc** [comment]
@lrwb-aou all changes made, I think this is in a good spot now, let me know your thoughts.

File: data_steward/tools/create_tier.py [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+    """
+    helper function to validate the tier parameter passed is either 'controlled' or 'registered'
+
+    :param tier: tier parameter passed through from either a list or command line argument
+    :return: nothing, breaks if not valid
+    """
+    if tier.lower() not in TIER_LIST:
+        LOGGER.error(
+            f"Parameter ERROR: {tier} is an incorrect input for the tier parameter, accepted: controlled or "
+            f"registered")
```
Comment (lrwb-aou) [inline data_steward/tools/create_tier.py:27]: Logging is nice.  Raising an error is even better.  We can't know which deid tier we are working with if this isn't provided.  Go ahead and raise an error.

File: data_steward/tools/create_tier.py [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+    """
+    helper function to validate the deid_stage parameter passed is correct, must be 'deid', 'base' or 'clean'
+
+    :param deid_stage: deid_stage parameter passed through from either a list or command line argument
+    :return: nothing, breaks if not valid
+    """
+    if deid_stage not in DEID_STAGE_LIST:
+        LOGGER.error(
+            f"Parameter ERROR: {deid_stage} is an incorrect input for the deid_stage parameter, accepted: "
+            f"deid, base, clean")
```
Comment (lrwb-aou) [inline data_steward/tools/create_tier.py:40]: Logging is nice.  Raising an error is even better.  We can't know which data_stage tier we are working with if this isn't provided.  Go ahead and raise an error.

File: data_steward/tools/create_tier.py [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+                        '--deid_stage',
+                        action='store',
+                        dest='deid_stage',
+                        help='deid stage (deid, base or clean)',
+                        required=True,
+                        choices=DEID_STAGE_LIST)
+    parser.add_argument('-l',
+                        '--console_log',
+                        dest='console_log',
+                        action='store_true',
```
Comment (lrwb-aou) [inline data_steward/tools/create_tier.py:128]: beautiful!

File: tests/unit_tests/data_steward/tools/create_tier_test.py [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+        invalid_deid_stage_params = [
+            'baseee', 'deid_base', 'deid_clean', 'clean_base'
+        ]
+
+        # Test if invalid parameters are given
+        for ds in invalid_deid_stage_params:
+            expected_error_output = f"ERROR:tools.create_tier:Parameter ERROR: {ds} is an incorrect input for " \
+                                    f"the deid_stage parameter, accepted: deid, base, clean"
+            with self.assertLogs() as cm:
+                validate_deid_stage_param(ds)
```
Comment (lrwb-aou) [inline tests/unit_tests/data_steward/tools/create_tier_test.py:165]: If you raise the error, you'll need to modify this test to catch it.

File: tests/unit_tests/data_steward/tools/create_tier_test.py [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+    def test_validate_tier_param(self):
+        # Preconditions
+        invalid_tier_params = ['foo', 'bar', 'controled', 'registry']
+
+        # Test if invalid parameters are given
+        for tier in invalid_tier_params:
+            expected_error_output = f"ERROR:tools.create_tier:Parameter ERROR: {tier} is an incorrect input for the " \
+                                    f"tier parameter, accepted: controlled or registered"
+            with self.assertLogs() as cm:
+                validate_tier_param(tier)
```
Comment (lrwb-aou) [inline tests/unit_tests/data_steward/tools/create_tier_test.py:151]: If you raise the error, modify this test to catch it.

**Review by lrwb-aou — CHANGES_REQUESTED** [review]
One more type of change and it'll be good to go.  Make it fail obnoxiously loudly.  :)

File: data_steward/tools/create_tier.py (hunk 15 above)
Comment (amcgrenera-vumc) [reply data_steward/tools/create_tier.py:41]: I've added in `raise argparse.ArgumentTypeError` I did leave in the `LOGGER.error` let me know if that is unnecessary and I should remove?

Comment (lrwb-aou) [reply data_steward/tools/create_tier.py:41]: I will not ask to remove logging statements!

File: data_steward/tools/create_tier.py [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+    """
+    helper function to validate the tier parameter passed is either 'controlled' or 'registered'
+
+    :param tier: tier parameter passed through from either a list or command line argument
+    :return: nothing, breaks if not valid
+    """
+    if tier.lower() not in TIER_LIST:
+        LOGGER.error(
+            f"Parameter ERROR: {tier} is an incorrect input for the tier parameter, accepted: controlled or "
+            f"registered")
```
Comment (lrwb-aou) [inline data_steward/tools/create_tier.py:27]: ok.  another thing, consider moving your message into a string variable, then you can send the message to the logger and raise the error.  For example,

``` 
if tier.lower() not in TIER_LIST:
    msg = "Bad argument". # your message is better, this is for example.
    LOGGER.error(msg)
    raise argparse.ArgumentTypeError(msg)
```

The empty `argparse.ArgumentTypeError` is functionally correct.  Adding the message will be more informative to whoever is running the script.

It's a difference of:
```
>>> import argparse
>>> raise argparse.ArgumentTypeError
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
argparse.ArgumentTypeError
>>> raise argparse.ArgumentTypeError('foo')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
argparse.ArgumentTypeError: foo
```
The second one would provide a more informative console message.

Comment (amcgrenera-vumc) [reply data_steward/tools/create_tier.py:27]: Awesome, thank you for this suggestion, much better console message.

**Review by lrwb-aou — APPROVED** [review]
Thank you!

**Pull request merged**